artichoke-backend/src/extn/core/enumerable/enumerable.rb

```diff
@@ -0,0 +1,450 @@
+##
+# Enumerable
+#
+module Enumerable
+
+  def drop(size)
+    size = size.__to_int
+    raise ArgumentError, "attempt to drop negative size" if size.negative?
+
+    ary = []
+    self.each {|*val| size.zero? ? ary << val.__svalue : size -= 1 }
+    ary
+  end
+
+  def drop_while(&block)
+    return to_enum :drop_while unless block
+
+    ary, state = [], false
+    self.each do |*val|
+      state = true if !state and !block.call(*val)
+      ary << val.__svalue if state
+    end
+    ary
+  end
+
+  def take(size)
+    size = size.__to_int
+    i = size.to_i
+    raise ArgumentError, "attempt to take negative size" if i.negative?
+    ary = []
+    return ary if i.zero?
+    self.each do |*val|
+      ary << val.__svalue
+      i -= 1
+      break if i.zero?
+    end
+    ary
+  end
+
+  def take_while(&block)
+    return to_enum :take_while unless block
+
+    ary = []
+    self.each do |*val|
+      return ary unless block.call(*val)
+      ary << val.__svalue
+    end
+    ary
+  end
+
+  def each_cons(size, &block)
+    size = size.__to_int
+    raise ArgumentError, "invalid size" if size <= 0
+
+    return to_enum(:each_cons,size) unless block
+    ary = []
+    size = size.to_i
+    self.each do |*val|
+      ary.shift if ary.size == size
+      ary << val.__svalue
+      block.call(ary.dup) if ary.size == size
+    end
+    nil
+  end
+
+  def each_slice(size, &block)
+    size = size.__to_int
+    raise ArgumentError, "invalid slice size" if size <= 0
+
+    return to_enum(:each_slice,size) unless block
+    ary = []
+    size = size.to_i
+    self.each do |*val|
+      ary << val.__svalue
+      if ary.size == size
+        block.call(ary)
+        ary = []
+      end
+    end
+    block.call(ary) unless ary.empty?
+    nil
+  end
+
+  def group_by(&block)
+    return to_enum :group_by unless block
+
+    h = {}
+    self.each do |*val|
+      key = block.call(*val)
+      sv = val.__svalue
+      h.key?(key) ? (h[key] << sv) : (h[key] = [sv])
+    end
+    h
+  end
+
+  def sort_by(&block)
+    return to_enum :sort_by unless block
+
+    ary = []
+    orig = []
+    self.each_with_index{|e, i|
+      orig.push(e)
+      ary.push([block.call(e), i])
+    }
+    if ary.size > 1
+      ary.sort!
+    end
+    ary.collect{|e,i| orig[i]}
+  end
+
+  def first(*args)
+    case args.length
+    when 0
+      self.each do |*val|
+        return val.__svalue
+      end
+      return nil
+    when 1
+      i = args[0].__to_int
```

Comment:
Methods that begin with an underscore (`_`) are mruby internal methods and are not part of Ruby Core. Can we reimplement this logic with methods in Ruby Core? Maybe `to_int` or `to_i`?